lib/saluki-components/src/sources/checks/mod.rs

```diff
@@ -57,82 +56,122 @@ pub struct ChecksConfiguration {
     #[serde(default)]
     additional_checksd: String,
 
-    /// Autodiscovery provider to use.
     #[serde(skip)]
-    autodiscovery_provider: Option<Arc<dyn AutodiscoveryProvider + Send + Sync>>,
+    env_provider: Option<E>,
+
+    #[serde(skip)]
+    full_configuration: Option<GenericConfiguration>,
 }
 
 impl ChecksConfiguration {
     /// Creates a new `ChecksConfiguration` from the given configuration.
     pub fn from_configuration(config: &GenericConfiguration) -> Result<Self, GenericError> {
-        Ok(config.as_typed()?)
+        let mut checks_config = config.as_typed::<Self>()?;
+        checks_config.full_configuration = Some(config.clone());
+        Ok(checks_config)
     }
+}
 
-    /// Sets the autodiscovery provider to use.
-    pub fn with_autodiscovery_provider<A>(mut self, provider: A) -> Self
-    where
-        A: AutodiscoveryProvider + Send + Sync + 'static,
-    {
-        self.autodiscovery_provider = Some(Arc::new(provider));
-        self
+impl<E> ChecksConfiguration<E> {
+    /// Sets the env provider to use.
+    pub fn with_env_provider<T>(self, env_provider: T) -> ChecksConfiguration<T> {
```

Comment:
Two things here:

- we try to prefer full words for _method_ names, and abbreviations/acronyms for variable names (just so the API documentation stays as human-friendly as possible)
- (this one is very very subjective) when following this sort of pattern, where you're replacing a value of type T (that implements a common trait) with another concrete type that implements the same common trait, we try to match the generic type parameter name but suffixed with a `2` (so `T2`), to sort of indicate that we're replacing the existing `T` with something `T`-like.. so in this case, our struct uses `E`, so we want to use `E2`

```suggestion
    /// Sets the environment provider to use.
    pub fn with_environment_provider<E2>(self, env_provider: E2) -> ChecksConfiguration<E2> {
```